Add example projects and tests

mrnkr opened the pull request:
This PR is what @bart-degreed had suggested in #2 , let me quote for reference:

> Add checks to the MongoDB repository that fail with an error whenever relationships are used. And add the example and test projects. But instead of commented-out tests, this would contain tests that assert an appropriate error is returned when using relationships. From what I read online, it looks like MongoDB requires IDs to always be strings. So we'll also need a test covering that. And there may be other corner cases to catch, such as the nullable type conversion. This PR would pin the current behavior, identifying what works and what does not.

Closes #1.

bart-degreed commented:
Please merge in latest master, I've setup cibuild there. And I created a v4-alpha1 release on GitHub, to test that it automatically publishes to NuGet.

I understand you don't want to make too many changes in this PR. The way I'm going to review the tests in this PR, is by running a file-by-file diff with the main repo. For each test, I'm going to ask myself questions like: Why are the assertions different here? Why was this test removed? What other cases should be added? Is this a duplicate? Is this testing serializer behavior, so it should be left out?

My point is that reviewing a large set with gaps (future work) is a lot harder than reviewing a set that is supposed to be complete. And once the set is complete, I'm going to ask you to remove several models, resource hooks etc, because they are unneeded.

I don't think this needs to result in a lot of changes on your side. For example, we have many tests for usage of the `?include=` query string parameter. You can substitute that whole set with a single test that fails whenever `?include=` is used. You'd just throw a `JsonApiException` with title "Relationships are not supported when using MongoDB." from the repository `ApplyQueryLayer` method whenever `QueryLayer.Include` is not null. Also throw when QueryLayer.Projection contains relationships. And there's a bit more to check, but you'll find them when adding such tests and I can help you how to fix that.

mrnkr commented:
I've just added error messages for when the user attempts to fetch a related resource (either via `?include=` or just `/primary/:id/secondary`) and for resource creation as well.

I also went back to the tests I'd said I was failing in #2 and I have some things to comment:
* Comparing attributes is indeed not supported by `MongoDB.Driver` https://jira.mongodb.org/browse/CSHARP-1592
* I left the test `Retrieves_all_properties_when_fieldset_contains_readonly_attribute` with a different name and asserting a `500` status code because in order to fix it I would need to actually implement `IModel.GetEntityTypes()` in `DummyModel`, right now it just throws `NotImplementedException`. Do you think there is a way for me to catch this case before it gets there so I can throw a proper error earlier?

As for resource updates I couldn't figure out what it would be like to test that because since you can't create how do you even think of updating, right? I understand we will most likely want to test this a bit anyway so some suggestions would come in handy!

bart-degreed commented:
> Retrieves_all_properties_when_fieldset_contains_readonly_attribute

To make the original test work, you just need to feed it the info `SelectClauseBuilder` is asking for. The code below replaces the static Dummy instance with an implementation that performs the lookup through the resource graph. It's mostly boilerplate from interface members.

<details>
<summary>MongoDbModel source code</summary>
<p>

```c#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using JsonApiDotNetCore.Configuration;
using Microsoft.EntityFrameworkCore.Infrastructure;
using Microsoft.EntityFrameworkCore.Metadata;

namespace JsonApiDotNetCore.MongoDb
{
    internal sealed class MongoDbModel : IModel
    {
        private readonly IResourceContextProvider _resourceContextProvider;

        public object this[string name] => throw new NotImplementedException();

        public MongoDbModel(IResourceContextProvider resourceContextProvider)
        {
            _resourceContextProvider = resourceContextProvider ?? throw new ArgumentNullException(nameof(resourceContextProvider));
        }

        public IEnumerable<IEntityType> GetEntityTypes()
        {
            var resourceContexts = _resourceContextProvider.GetResourceContexts();
            return resourceContexts.Select(resourceContext => new MongoEntityType(resourceContext, this));
        }

        public IAnnotation FindAnnotation(string name) => throw new NotImplementedException();
        public IEnumerable<IAnnotation> GetAnnotations() => throw new NotImplementedException();
        public IEntityType FindEntityType(string name) => throw new NotImplementedException();
        public IEntityType FindEntityType(string name, string definingNavigationName, IEntityType definingEntityType) => throw new NotImplementedException();
    }

    internal sealed class MongoEntityType : IEntityType
    {
        private readonly ResourceContext _resourceContext;

        public IModel Model { get; }
        public Type ClrType => _resourceContext.ResourceType;

        public string Name => throw new NotImplementedException();
        public IEntityType BaseType => throw new NotImplementedException();
        public string DefiningNavigationName => throw new NotImplementedException();
        public IEntityType DefiningEntityType => throw new NotImplementedException();
        public object this[string name] => throw new NotImplementedException();

        public MongoEntityType(ResourceContext resourceContext, MongoDbModel owner)
        {
            _resourceContext = resourceContext ?? throw new ArgumentNullException(nameof(resourceContext));
            Model = owner ?? throw new ArgumentNullException(nameof(owner));
        }

        public IEnumerable<IProperty> GetProperties()
        {
            return _resourceContext.Attributes.Select(attr => new MongoDbProperty(attr.Property, this));
        }

        public IAnnotation FindAnnotation(string name) => throw new NotImplementedException();
        public IEnumerable<IAnnotation> GetAnnotations() => throw new NotImplementedException();
        public IKey FindPrimaryKey() => throw new NotImplementedException();
        public IKey FindKey(IReadOnlyList<IProperty> properties) => throw new NotImplementedException();
        public IEnumerable<IKey> GetKeys() => throw new NotImplementedException();
        public IForeignKey FindForeignKey(IReadOnlyList<IProperty> properties, IKey principalKey, IEntityType principalEntityType) => throw new NotImplementedException();
        public IEnumerable<IForeignKey> GetForeignKeys() => throw new NotImplementedException();
        public IIndex FindIndex(IReadOnlyList<IProperty> properties) => throw new NotImplementedException();
        public IEnumerable<IIndex> GetIndexes() => throw new NotImplementedException();
        public IProperty FindProperty(string name) => throw new NotImplementedException();
        public IServiceProperty FindServiceProperty(string name) => throw new NotImplementedException();
        public IEnumerable<IServiceProperty> GetServiceProperties() => throw new NotImplementedException();
    }

    internal sealed class MongoDbProperty : IProperty
    {
        public IEntityType DeclaringEntityType { get; }
        public PropertyInfo PropertyInfo { get; }

        public string Name => throw new NotImplementedException();
        public Type ClrType => throw new NotImplementedException();
        public FieldInfo FieldInfo => throw new NotImplementedException();
        public ITypeBase DeclaringType => throw new NotImplementedException();
        public bool IsNullable => throw new NotImplementedException();
        public ValueGenerated ValueGenerated => throw new NotImplementedException();
        public bool IsConcurrencyToken => throw new NotImplementedException();
        public object this[string name] => throw new NotImplementedException();

        public MongoDbProperty(PropertyInfo propertyInfo, MongoEntityType owner)
        {
            DeclaringEntityType = owner ?? throw new ArgumentNullException(nameof(owner));
            PropertyInfo = propertyInfo ?? throw new ArgumentNullException(nameof(propertyInfo));
        }

        public IAnnotation FindAnnotation(string name) => throw new NotImplementedException();
        public IEnumerable<IAnnotation> GetAnnotations() => throw new NotImplementedException();
    }
}
```

</p>
</details> 

Please put each of these classes in its own file. We allow multiple types in a single file when nested or only their generic arguments vary.

bart-degreed commented:
> As for resource updates I couldn't figure out what it would be like to test that because since you can't create how do you even think of updating, right? I understand we will most likely want to test this a bit anyway so some suggestions would come in handy!

The use case is that an API developer can define resources with `HasOne`/`HasMany`/`HasManyThrough` attributes on them. And then his/her customer can send a GET/POST/PATCH that refers to those relationships.

My first thought to block relationships was to just forbid using these attributes in resources. This repo could add a post-processor step in the resource graph building process that throws when relationships are modelled. But that is too restrictive, because an API may serve both EF Core and MongoDB resources at the same time.

The only place where we run into trouble, is when the code enters `MongoDbRepository`, so that is the place to intercept.

So to answer your question: A test that shows PATCH fails when relationships are being sent in the request body would not actually need to create a resource with a relationship in the test-database upfront (there is currently no way to do that). Not a problem, because the API code never gets there. But the resource class being used in the test _does_ need to define a relationship. Then the code passes through query string parsing without errors, ending up in the repo.

bart-degreed commented:
I've taken a peek at your changes in `MongoDbRepository` and couldn't resist trying to help a bit to get it right. Doing so requires deep knowledge of the various expressions and how they compose. Since I've built that and its still fresh in my memory I was able to come up with the validator below within an hour. I Hope I didn't spoil the fun for you, please let me know if you want me to hold back a bit more.

<details>
<summary>MongoDbQueryExpressionValidator source code</summary>
<p>

```c#
internal sealed class MongoDbQueryExpressionValidator : QueryExpressionRewriter<object>
{
    public void Validate(QueryLayer layer)
    {
        if (layer == null) throw new ArgumentNullException(nameof(layer));

        bool hasIncludes = layer.Include?.Elements.Any() == true;
        var hasSparseRelationshipSets = layer.Projection?.Any(pair => pair.Key is RelationshipAttribute) == true;

        if (hasIncludes || hasSparseRelationshipSets)
        {
            throw CreateErrorForRelationshipUsage();
        }

        ValidateExpression(layer.Filter);
        ValidateExpression(layer.Sort);
        ValidateExpression(layer.Pagination);
    }

    private static JsonApiException CreateErrorForRelationshipUsage()
    {
        return new JsonApiException(new Error(HttpStatusCode.BadRequest)
        {
            Title = "Relationships are not supported when using MongoDB."
        });
    }

    private void ValidateExpression(QueryExpression expression)
    {
        if (expression != null)
        {
            Visit(expression, null);
        }
    }

    public override QueryExpression VisitResourceFieldChain(ResourceFieldChainExpression expression, object argument)
    {
        if (expression != null)
        {
            if (expression.Fields.Count > 1 || expression.Fields.First() is RelationshipAttribute)
            {
                throw CreateErrorForRelationshipUsage();
            }
        }

        return base.VisitResourceFieldChain(expression, argument);
    }

    public override QueryExpression VisitComparison(ComparisonExpression expression, object argument)
    {
        if (expression?.Left is ResourceFieldChainExpression && expression.Right is ResourceFieldChainExpression)
        {
            throw new JsonApiException(new Error(HttpStatusCode.BadRequest)
            {
                Title = "Comparing attributes against each other is not supported when using MongoDB."
            });
        }

        return base.VisitComparison(expression, argument);
    }
}
```

</p>
</details> 

You can call `Validate()` from `MongoDbRepository.ApplyQueryLayer`, just after its parameter null assertion. Then you can remove your custom catch block in `GetAsync`, `AssertNoInclude` etc. 

I noticed you were adding the "include" parameter to the exception, which is fine for include, but hard to reconstruct for other query string parameters such as fields[]. So I took that out, it is not important enough to do extra work for it.

Also I created a private method to produce the exception when a relationship is encountered. You'll likely also need that from within `MongoDbRepository`, so I think its best to create a custom exception for that. See the exceptions in JsonApiDotNetCore.Errors namespace in the main repo for inspiration.

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/Internal/MongoDbModel.cs:
```diff
@@ -0,0 +1,32 @@
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Metadata;
+
+namespace JsonApiDotNetCore.MongoDb.Internal
```
Top-level Internal namespaces are an anti-pattern, similar to Shared, Common, Utilities etc. because they tend to be filled with unrelated types (which defeats the purpose of having namespaces). Nested Internal namespaces are reserved for [pubternal](https://amp.reddit.com/r/dotnet/comments/jwzgl9/word_of_the_day_pubternal/) types, which these are not. Instead I recommend to put these, along with MongoDbRepository in JsonApiDotNetCore.MongoDb.Repositories.

bart-degreed replied:
> Instead I recommend to put these, along with MongoDbRepository in JsonApiDotNetCore.MongoDb.Repositories

Did you intentionally place them in the project root instead?

You've placed `MongoDbQueryExpressionValidator` in `JsonApiDotNetCore.MongoDb.Queries.Expressions`, which is the namespace where immutable expression objects live. This is not one of those, instead it is validation logic that is closely related to the repository and not a generic building block, so I would put it there too.

bart-degreed replied:
Also, please move GettingStarted and Example projects inside Examples subfolder, similar to the main repo.

mrnkr replied:
In my last commit I did put all these classes in `JsonApiDotNetCore.MongoDb.Repositories`, sorry about that!

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/Errors/UnsupportedRelationshipException.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace JsonApiDotNetCore.MongoDb.Errors
+{
+    public sealed class UnsupportedRelationshipException : JsonApiException
```
Please add a summary here too.

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoDbRepository.cs:
```diff
@@ -50,20 +48,12 @@ public class MongoDbRepository<TResource, TId>
         {
             if (layer == null) throw new ArgumentNullException(nameof(layer));
+            
+            var queryExpressionValidator = new MongoDbQueryExpressionValidator();
```
This method calls ApplyQueryLayer, which performs the same so can be removed here.

bart-degreed commented on test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/Filtering/FilterOperatorTests.cs:
```diff
@@ -98,8 +98,7 @@ public async Task Cannot_filter_equality_on_two_attributes()
             responseDocument.Errors[0].StatusCode.Should().Be(HttpStatusCode.BadRequest);
-            responseDocument.Errors[0].Title.Should().Be("MongoDB does not allow comparing two fields to each other, only constants.");
-            responseDocument.Errors[0].Detail.Should().Be("Unsupported filter: ({document}{SomeInt32} == {document}{OtherInt32}).");
+            responseDocument.Errors[0].Title.Should().Be("Comparing attributes against each other is not supported when using MongoDB.");
```
Should also assert that .Detail is null. When that changes in the future, this breaks the test so we know about it.

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoDbRepository.cs:
```diff
@@ -153,10 +130,7 @@ private void AssertNoRelationship(TResource resourceFromRequest)
-                    {
-                        Title = "Relationships are not supported when using MongoDB."
-                    });
+                    throw new UnsupportedRelationshipException();
```
It is sufficient to throw just when `_targetedFields.Relationships.Any()`

Better name for this method: AssertNoRelationshipsAreTargeted. And method parameter can be removed.

bart-degreed reviewed (changes requested):
Review of your last commit only. Please let me know when you're ready for review of the rest.

mrnkr commented:
Please don't hold back on my account! Thank you so much for those snippets, they worked great! Knowing the codebase does help when coming up with solutions to these problems hehe.

Anyway, I just added a test for updating relationships, cleaned up the models that had to have relationships and addressed the comments of your review. I think I'm ready for a review of the entire PR.

bart-degreed commented on test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/Meta/TopLevelCountTests.cs:
```diff
@@ -0,0 +1,127 @@
+                }
+            };
+
+            var route = $"/api/v1/todoItems/{todoItem.Id}";
```
Should be `todoItem.StringId`. Why is this different from main repo?

bart-degreed reviewed (changes requested):
Reviewed until Meta tests, will look at the rest later.

bart-degreed commented:
@mrnkr You mentioned [here](https://github.com/json-api-dotnet/JsonApiDotNetCore.MongoDb/pull/2#issuecomment-751124662) that resource metadata is broken. Can you elaborate on that?

mrnkr commented:
> @mrnkr You mentioned [here](https://github.com/json-api-dotnet/JsonApiDotNetCore.MongoDb/pull/2#issuecomment-751124662) that resource metadata is broken. Can you elaborate on that?

Yes, I'd forgotten about that... The test I was failing was `ResourceMetaTests.ResourceDefinition_That_Implements_GetMeta_Contains_Resource_Meta` with the following exception:

```
JsonApiDotNetCore.MongoDb.Example.Tests.IntegrationTests.Meta.ResourceMetaTests.ResourceDefinition_That_Implements_GetMeta_Contains_Resource_Meta

Xunit.Sdk.XunitException
Expected responseDocument.ManyData[0].Meta to contain keys {"hasHighPriority"}, but found <null>.
   at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Collections.GenericDictionaryAssertions`2.ContainKeys(IEnumerable`1 expected, String because, Object[] becauseArgs)
   at FluentAssertions.Collections.GenericDictionaryAssertions`2.ContainKey(TKey expected, String because, Object[] becauseArgs)
   at JsonApiDotNetCore.MongoDb.Example.Tests.IntegrationTests.Meta.ResourceMetaTests.ResourceDefinition_That_Implements_GetMeta_Contains_Resource_Meta() in /Users/mrnkr/Documents/code/JsonApiDotNetCore.MongoDb/test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/Meta/ResourceMetaTests.cs:line 53
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
--- End of stack trace from previous location where exception was thrown ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

Basically, the metadata just isn't added to the response object.

mrnkr commented on test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/Filtering/FilterOperatorTests.cs:
```diff
@@ -0,0 +1,409 @@
+
+            var options = (JsonApiOptions) _testContext.Factory.Services.GetRequiredService<IJsonApiOptions>();
+            options.EnableLegacyFilterNotation = false;
+            options.SerializerSettings.DateFormatString = "yyyy-MM-dd";
```
Dunno if this is the only solution but it's the only way I could work around tests failing because of the timezone I'm at.

```
JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.Filtering.FilterOperatorTests.Can_filter_comparison_on_DateTime.Can_filter_comparison_on_DateTime(matchingDateTime: "2001-01-01", nonMatchingDateTime: "2001-01-09", filterOperator: LessOrEqual, filterDateTime: "2001-01-05")

Xunit.Sdk.XunitException
Expected responseDocument.ManyData[0].Attributes["someDateTime"] to be "2001-01-01", but found <2001-01-01 03:00:00>.
   at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Primitives.ObjectAssertions.Be(Object expected, String because, Object[] becauseArgs)
   at JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.Filtering.FilterOperatorTests.Can_filter_comparison_on_DateTime(String matchingDateTime, String nonMatchingDateTime, ComparisonOperator filterOperator, String filterDateTime) in /Users/mrnkr/Documents/code/JsonApiDotNetCore.MongoDb/test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Filtering/FilterOperatorTests.cs:line 226
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
--- End of stack trace from previous location where exception was thrown ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Filtering/FilterOperatorTests.cs:
```diff
@@ -0,0 +1,410 @@
+
+            var options = (JsonApiOptions) _testContext.Factory.Services.GetRequiredService<IJsonApiOptions>();
+            options.EnableLegacyFilterNotation = false;
+            options.SerializerSettings.DateFormatString = "yyyy-MM-dd";
```
> Dunno if this is the only solution but it's the only way I could work around tests failing because of the timezone I'm at.

The tests are supposed to cover our important use cases. When such a test fails, it provides an opportunity to explore what is going on. Is the test wrong, is it hitting a bug, can it be implemented differently or is this supposed to fail?

mrnkr replied:
I explained myself poorly last time, I said it was because of the timezone I was at but it's not exactly that.

When I remove line 38 I fail the test `Can_filter_comparison_on_DateTime` because the date value that is returned by the API has the time included in it but the expected value does not. Here's an example of a failed assertion:

```
Expected responseDocument.ManyData[0].Attributes["someDateTime"] to be <2001-01-01>, but found <2001-01-01 03:00:00>.
```

Personally I don't think this is a bug, I don't remember how postgres handles dates and whether this can have anything to do with mongo storing dates differently but it might be something along those lines...

bart-degreed reviewed (commented):
I've gone over the open feedback items today, and noticed that some of them aren't yet addressed, others only partly.

So I'd like to structure our way of working a bit, as it is becoming unclear to me what is going on. Let's follow these steps:
1. When you're ready for another review round, use the GitHub button to re-request review
2. While reviewing, I'm adding new feedback items
3. Once you've addressed a feedback item, mark it with a thumbs-up -or- comment so we can discuss
4. I'll verify and mark addressed feedback items as resolved

mrnkr commented on test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/Filtering/FilterDataTypeTests.cs:
```diff
@@ -0,0 +1,338 @@
+
+            var options = (JsonApiOptions) _testContext.Factory.Services.GetRequiredService<IJsonApiOptions>();
+            options.EnableLegacyFilterNotation = false;
+            options.SerializerSettings.DateTimeZoneHandling = DateTimeZoneHandling.Utc;
```
It wasn't, I'd left it there by mistake so I removed it. I must have had a problem with timezones here too.

mrnkr commented on src/JsonApiDotNetCore.MongoDb.Example/Models/User.cs:
```diff
@@ -0,0 +1,40 @@
+        [Attr]
+        public string Id { get; set; }
+        
+        // private readonly ISystemClock _systemClock;
```
There is one test that uses this resource but I'm not sure I need to keep it... It is `SparseFieldSetTests.Cannot_select_attribute_with_blocked_capability`.

It does seem to be testing behavior I didn't modify, if you can confirm this then I can remove the test and the resource altogether.

In any case, all that is needed is for the password to be blocked for that test so having the clock does not provide us with anything IMO.

bart-degreed replied:
If you search the main repo for "Retrieving the requested attribute is not allowed.", you'll find only SparseFieldSetQueryStringParameterReader.cs, which performs query string parsing. MongoDB does not change that, so the test can be removed.

mrnkr replied:
I removed only that test and the `User` class. Let me know if you'd meant me to remove the entire test class.

mrnkr commented on test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/Meta/TopLevelCountTests.cs:
```diff
@@ -0,0 +1,127 @@
+using MongoDB.Driver;
+using Xunit;
+
+namespace JsonApiDotNetCore.MongoDb.Example.Tests.IntegrationTests.Meta
```
Sorry about this one, I was trying to undo some changes and ended up uploading this old version by mistake, I removed it now.

mrnkr commented on test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/Filtering/FilterDepthTests.cs:
```diff
@@ -0,0 +1,87 @@
+
+namespace JsonApiDotNetCore.MongoDb.Example.Tests.IntegrationTests.Filtering
+{
+    public sealed class FilterDepthTests : IClassFixture<IntegrationTestContext<Startup>>
```
I did call these `Cannot_<rest_of_name>` since that's what they're actually testing but I did add them. Thought it better to let you know.

mrnkr commented:
Sorry things got a bit confusing... I think I did get all your comments sorted out this time, if not I did leave comments.

You'll notice I'm a bit lost in terms of how to proceed in terms of the `options.SerializerSettings.DateFormatString` issue and resource metadata. Aside from that I think I'm ready for the next review round.

bart-degreed commented on src/GettingStarted/Startup.cs:
```diff
@@ -0,0 +1,52 @@
+                return client.GetDatabase(Configuration.GetSection("DatabaseSettings:Database").Value);
+            });
+
+            services.AddResourceRepository<MongoDbRepository<Book>>();
```
This is still reversed in README.md

mrnkr replied:
Sorry I'd missed that one, I've just set it right.

bart-degreed commented:
> Dunno if this is the only solution but it's the only way I could work around tests failing because of the timezone I'm at.

I investigated, turns out that MongoDB always stores in UTC zone. Pushed a commit to fix this.

bart-degreed commented:
>> @mrnkr You mentioned here that resource metadata is broken. Can you elaborate on that?

> Yes, I'd forgotten about that... The test I was failing was ResourceMetaTests.ResourceDefinition_That_Implements_GetMeta_Contains_Resource_Meta with the following exception

This failed because MongoDbExample is not using service discovery (the example in main repo does), so resource definitions are not registered automatically so you need to do that yourself. I've pushed a commit to add the working test.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Pagination/PaginationWithTotalCountTests.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        [Fact]
+        public async Task Cannot_paginate_in_single_primary_resource()
```
This test validates query string parsing, so can be removed.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Pagination/PaginationWithTotalCountTests.cs:
```diff
@@ -0,0 +1,105 @@
+            responseDocument.Errors[0].Detail.Should().Be("This query string parameter can only be used on a collection of resources (not on a single resource).");
+            responseDocument.Errors[0].Source.Parameter.Should().Be("page[number]");
+        }
+    }
```
The following tests are missing, which I think should be added:
- Uses_default_page_number_and_size: but on a primary resource instead (/api/v1/articles)
- Returns_all_resources_when_paging_is_disabled: but on a primary resource instead (/api/v1/articles)

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Pagination/RangeValidationTests.cs:
```diff
@@ -0,0 +1,152 @@
+
+namespace JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.Pagination
+{
+    public sealed class RangeValidationTests : IClassFixture<IntegrationTestContext<Startup>>
```
All tests, _except_ the following, can be removed, because they test query string validation logic.
- When_page_number_is_too_high_it_must_return_empty_set_of_resources
- When_page_size_is_zero_it_must_succeed
- When_page_size_is_positive_it_must_succeed

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Pagination/RangeValidationWithMaximumTests.cs:
```diff
@@ -0,0 +1,144 @@
+
+namespace JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.Pagination
+{
+    public sealed class RangeValidationWithMaximumTests : IClassFixture<IntegrationTestContext<Startup>>
```
These tests can be removed, they verify paging query strings against configured options.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/QueryStrings/QueryStringTests.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.QueryStrings
+{
+    public sealed class QueryStringTests : IClassFixture<IntegrationTestContext<Startup>>
```
These tests can be removed, they verify query string parsing.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,202 @@
+using MongoDB.Driver;
+using Xunit;
+using Person = JsonApiDotNetCoreMongoDbExample.Models.Person;
+
```
Please also add all tests from ResourceDefinitionQueryCallbackTests.cs, except `Include_from_resource_definition_has_blocked_capability`.

These influence the query being sent to the database, so it hits MongoDb repository logic.

mrnkr replied:
I'm still failing the two tests that filter by resource metadata so this one is to be considered partially addressed too.

I'm referring to the following tests:
* Queryable_parameter_handler_from_resource_definition_and_query_string_filter_are_applied
* Queryable_parameter_handler_from_resource_definition_is_applied

bart-degreed replied:
They fail because the following needs to be added: https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/master/src/JsonApiDotNetCore/Repositories/EntityFrameworkCoreRepository.cs#L93-L103

mrnkr replied:
Looked at `EntityFrameworkCoreRepository` in the original repo and figured out what was missing, just added a commit adding support for custom query parameters so this one is finally addressed.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,202 @@
+        }
+
+        [Fact]
+        public async Task Cannot_sort_in_single_primary_resource()
```
This test concerns query string validation, can be removed.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,202 @@
+        }
+
+        [Fact]
+        public async Task Cannot_sort_in_single_secondary_resource()
```
Same here: query string validation, so can be removed.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,202 @@
+        }
+
+        [Fact]
+        public async Task Cannot_sort_on_attribute_with_blocked_capability()
```
This test can be removed.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,202 @@
+            responseDocument.Errors[0].Detail.Should().Be("This query string parameter can only be used on a collection of resources (not on a single resource).");
+            responseDocument.Errors[0].Source.Parameter.Should().Be("sort");
+        }
+
```
Add the following missing tests:
- Can_sort_on_HasMany_relationship
- Can_sort_on_HasManyThrough_relationship
- Can_sort_on_HasOne_relationship

The difference is they are supposed to fail (assert on that)

mrnkr replied:
I'm still failing `Cannot_sort_on_HasMany_relationship` so consider this partially address for now.

bart-degreed replied:
This test needs a bit more to bypass an optimization. The MongoDB collection is empty, so when `.CountAsync()` is called, no documents are returned, so the actual query to fetch data never executes, so the sort is never applied. The solution is to add a single document upfront and the test passes.

mrnkr replied:
I had copied these tests without the complete `Arrange` part, I mean, I had not added resources to the database prior to calling the endpoints. This worked for `HasOne` and `HasManyThrough` but it did not work for `HasMany`. Adding this part of the test made `HasMany` work too. Do you think we should do something about this?

I mean, should we expect an API to return `400` when the database is empty and we try to use relationships? I think we should but I don't know how I can do that neatly hehe

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,202 @@
+            // Assert
+            httpResponse.Should().HaveStatusCode(HttpStatusCode.OK);
+
+            persons.Sort((a, b) => a.LastName.CompareTo(b.LastName) + b.Id.CompareTo(a.Id));
```
Instead of this, take the original test and set hardcoded object IDs (you can get some by calling `ObjectId.GenerateNewId()`). The test needs to show it actually performs the sorting, irrespective of the order of insertion. And show it applies nested sorting.

This applies to test `Sorts_by_ID_if_none_specified` too.

mrnkr replied:
I did do this but I want to point out that sorting by ID when working with MongoDB will always like sorting by order of insertion since an `ObjectId` is generated based on the current timestamp, a random number and an auto-incrementing counter. Here is the official reference: https://docs.mongodb.com/manual/reference/method/ObjectId/

It is for this reason that in `Can_sort_descending_by_ID` you'll notice I first generate the three ids sequentially and then assign them in reverse order to the `Persons`.

bart-degreed replied:
> I did do this but I want to point out that sorting by ID when working with MongoDB will always like sorting by order of insertion since an ObjectId is generated based on...

I'm aware of that. And it is the reason I'd like to use _hardcoded_ IDs here. This emulates an API client sending us explicit IDs instead of the server auto-generating them, so they can be in a different order.

To clarify: don't call `ObjectId.GenerateNewId()` in the test code, but execute it in your terminal and copy/paste the resulting values.

bart-degreed replied:
Oh, and `persons` should be `people`.

bart-degreed replied:
I don't see any changes.

mrnkr commented:
Thank you so much for your help! I could not have come up with those solutions myself.

I've just addressed the two remaining feedback items, let me know what else needs to be done when you can.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/SparseFieldSets/SparseFieldSetTests.cs:
```diff
@@ -0,0 +1,273 @@
+                await collection.InsertOneAsync(article);
+            });
+
+            var route = "/api/v1/articles?fields[articles]=caption"; // TODO: once relationships are implemented select author field too
```
The original tests are:
- Can_select_fields_in_primary_resources: one attribute + one relationship
- Can_select_attribute_in_primary_resources: only one attribute
- Can_select_relationship_in_primary_resources: only one relationship
- ...
- Can_select_fields_in_primary_resource_by_ID

So I think these should become:
- Cannot_select_fields_with_relationship_in_primary_resources
- Can_select_attribute_in_primary_resources
- Cannot_select_relationship_in_primary_resources
- ...
- Can_select_attribute_in_primary_resource_by_ID

Also add negatives for Can_select_fields_of_HasOne_relationship, Can_select_fields_of_HasMany_relationship, Can_select_fields_of_HasManyThrough_relationship.

Remove these (query string validations):
- Cannot_select_on_unknown_resource_type
- Cannot_select_attribute_with_blocked_capability

mrnkr replied:
For now I've only removed the tests you said to remove. 

I don't quite understand what I should do about the rest... I mean, I do understand having to add negatives for relationships but not the rest. Can you explain this?

bart-degreed replied:
The way I look at the tests is by comparing them against the main repo. If you want to see what I see, use [Araxis Merge](https://www.araxis.com/merge/index.en), which is the best diff/merge tool I've ever encountered. There's a 30-day trial available.

So what I'm proposing is to take everything from main, then apply the mentioned changes. So (...) means: unchanged. Hope this clarifies my intentions.

mrnkr replied:
Right now I've just added the tests for relationships.

mrnkr replied:
@bart-degreed I was just trying to bring in the `Cannot_select_fields_with_relationship_in_primary_resources` test and it turns out that the url `/api/v1/articles?fields[articles]=caption,author` does not fail... The `QueryLayer` has a projection with only two fields, them being `id` and `caption` but no `author` in sight so the validator cannot do its job here. I checked that `Author` was declared as a `HasOne` relationship and it is. Any ideas?

bart-degreed replied:
Hmm... yes, this is correct. Because there is no `include`, the relationship is not fetched. However, putting `author` in the fieldset results in rendering the relationship link (without any data in it).

See the note "Pay attention to the fact that you have to add..." at https://jsonapi.org/examples

I haven't tried, but for your test to fail, I think you need to add the following code to `MongoDbQueryExpressionValidator`:
```c#
public override QueryExpression VisitSparseFieldSet(SparseFieldSetExpression expression, object argument)
{
    if (expression != null && expression.Fields.Any(field => field is RelationshipAttribute))
    {
        throw new UnsupportedRelationshipException();
    }

    return base.VisitSparseFieldSet(expression, argument);
}
```

bart-degreed replied:
Oh well, that does not work, because relationships in sparse fieldsets do not propagate up till there. Instead, add the following to `MongoDbRepository.ApplyQueryLayer`:
```c#
bool hasRelationshipSelectors = _constraintProviders
    .SelectMany(p => p.GetConstraints())
    .Select(expressionInScope => expressionInScope.Expression)
    .OfType<SparseFieldTableExpression>()
    .Any(fieldTable => fieldTable.Table.Values.Any(fieldSet => fieldSet.Fields.Any(field => field is RelationshipAttribute)));

if (hasRelationshipSelectors)
{
    throw new UnsupportedRelationshipException();
}
```

mrnkr replied:
Worked like a charm! Thank you 😊

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/SparseFieldSets/SparseFieldSetTests.cs:
```diff
@@ -0,0 +1,273 @@
+            responseDocument.ManyData[0].Id.Should().Be(article.StringId);
+            responseDocument.ManyData[0].Attributes.Should().HaveCount(1);
+            responseDocument.ManyData[0].Attributes["caption"].Should().Be(article.Caption);
+
```
Undo the removal of relationship assertions in this file. Do assert they are null.

bart-degreed reviewed (changes requested):
Gone though all code, except for the tests in ReadWrite. Will look at those later.

bart-degreed commented:
It would be nice if you take another pass over the ReadWrite tests, based on recent insights. For example, tests like these (validations at unrelated layers in the architecture) should be removed:
- Sets_location_header_for_created_resource
- Cannot_create_resource_with_client_generated_ID
- Cannot_create_resource_for_missing_request_body
- Cannot_create_resource_for_missing_type
- Cannot_create_resource_for_unknown_type
- Cannot_create_resource_on_unknown_resource_type_in_url
- Cannot_create_on_resource_type_mismatch_between_url_and_body
- Cannot_create_resource_attribute_with_blocked_capability
- Cannot_create_resource_for_broken_JSON_request_body

mrnkr commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTestContext.cs:
```diff
@@ -0,0 +1,227 @@
+                        webBuilder.UseStartup<TStartup>();
+
+                        webBuilder.ConfigureTestServices(services =>
+                        {
```
Not that I can remember, I looked at the implementation of `IntegrationTestContext` in the original repo and left it like it was there.

bart-degreed replied:
That's strange. Looking at the [diff](https://github.com/json-api-dotnet/JsonApiDotNetCore/commits/master/test/JsonApiDotNetCoreExampleTests/IntegrationTestContext.cs), it has always been `ConfigureServices` since I added this file last summer.

mrnkr commented:
In my last commit I tried to remove all the tests that were unnecessary and address as many of the comments but I didn't get to address all of them yet.

I left a question about one of the feedback items, let you know about the three tests I'm still failing and I still have to check that all tests I have in the `ReadWrite` namespace actually make sense.

mrnkr commented:
With my latest commits I think I was able to address almost everything, you'll see I left one more comment about the only test I'm not able to get right at the moment.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Filtering/FilterDepthTests.cs:
```diff
@@ -0,0 +1,224 @@
+                var collection = db.GetCollection<Article>(nameof(Article));
+                await collection.DeleteManyAsync(Builders<Article>.Filter.Empty);
+                await collection.InsertManyAsync(articles);
+            });
```
> to the root of the **_test_** project, so you can write:

Why did you add this to the library? It is only used for convenience to write tests a bit shorter.

mrnkr replied:
My bad, this was just a distraction on my end

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateResourceTests.cs:
```diff
@@ -0,0 +1,502 @@
+        {
+            _testContext = testContext;
+            
+            _testContext.RegisterResources(builder =>
+            {
+                builder.Add<RgbColor, string>();
+                builder.Add<UserAccount, string>();
+                builder.Add<WorkItem, string>();
+            });
```
You can get rid of these everywhere by using auto-discovery. To do that:
1. In JsonApiDotNetCoreMongoDbExample.Startup, replace:
```c#
resources: builder => { .. }
```
with:
```c#
discovery: facade => facade.AddCurrentAssembly()
```
2. In IntegrationTestContext.CreateFactory, replace ```resources: _registerResources``` with the same. Next, you can remove the _registerResources callback and its method.
3. You _could_ do the same for the GettingStarted project, but I think its nice to keep that one as-is, so we demonstrate both possibilities.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateResourceTests.cs:
```diff
@@ -0,0 +1,502 @@
+                builder.Add<WorkItem, string>();
+            });
+            
+            _testContext.ConfigureServicesAfterStartup(services =>
+            {
+                services.AddResourceRepository<MongoDbRepository<RgbColor>>();
+                services.AddResourceRepository<MongoDbRepository<UserAccount>>();
+                services.AddResourceRepository<MongoDbRepository<WorkItem>>();
+            });
```
Knowing that the example project exclusively uses MongoDB (not a mixed form with EF Core), it is possible to overwrite the built-in open-generic registrations for `EntityFrameworkCoreRepository` with registrations for `MongoDbRepository`, so these calls can be removed.

To do that:
1. In JsonApiDotNetCoreMongoDbExample.Startup, replace the lines with ```services.AddResourceRepository``` with the following:
```c#
services.AddScoped(typeof(IResourceReadRepository<,>), typeof(MongoDbRepository<,>));
services.AddScoped(typeof(IResourceWriteRepository<,>), typeof(MongoDbRepository<,>));
services.AddScoped(typeof(IResourceRepository<,>), typeof(MongoDbRepository<,>));
```
2. Add the same to TestableStartup.ConfigureServices.
3. Likewise, I'd prefer to keep the GettingStarted project as-is.

bart-degreed reviewed (changes requested):
Haven't looked at the majority of the ReadWrite tests yet, but given some pointers for cleanup.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,219 @@
+        }
+        
+        [Fact]
+        public async Task Cannot_sort_on_HasManyThrough_relationship()
```
> I had copied these tests without the complete Arrange part, I mean, I had not added resources to the database prior to calling the endpoints. This worked for HasOne and HasManyThrough but it did not work for HasMany. Adding this part of the test made HasMany work too. Do you think we should do something about this?
I mean, should we expect an API to return 400 when the database is empty and we try to use relationships? I think we should but I don't know how I can do that neatly hehe

In theory we should, however that involves complexity which I believe is not worth the effort (plus extra setup burden for end-users). The reason some fail and other don't worries me. I suspect it has to do with the order of tests execution. So a previous test inserted a record that is reused. To verify, clear the whole collection upfront. I believe it's best to insert a record upfront for all three relationship tests in order to get stable results. We don't want tests to break when they are reordered or a new one is inserted.

mrnkr replied:
Agreed but wait, you mean some of these tests are failing on your end?

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Sorting/SortTests.cs:
```diff
@@ -0,0 +1,219 @@
+
+            await _testContext.RunOnDatabaseAsync(async db =>
+            {
+                await db.ClearCollectionAsync<Blog>();
```
The point is to have at least one record, so clearing is not needed.

mrnkr commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/Pagination/PaginationWithTotalCountTests.cs:
```diff
@@ -0,0 +1,165 @@
+    {
+        private const int _defaultPageSize = 5;
+        private readonly IntegrationTestContext<Startup> _testContext;
+        private readonly Faker<TodoItem> _todoItemFaker = new Faker<TodoItem>();
```
I'm actually using IntelliJ Rider so I have Resharper, I'll pay more attention to this and try to get as many of these unused variables out.

mrnkr commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/SparseFieldSets/SparseFieldSetTests.cs:
```diff
@@ -0,0 +1,384 @@
+        }
+        
+        [Fact]
+        public async Task Can_select_fields_of_HasManyThrough_relationship()
```
Although I think I should get this by now I don't so... Why could it be that as soon as I removed the `include` the test started failing again?

mrnkr commented:
Left the new tests for `ReadWrite` for tomorrow and left a couple questions.

bart-degreed reviewed (changes requested):
Gone over open items, think I addressed all your questions.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Creating/CreateResourceWithRelationshipsTests.cs:
```diff
@@ -0,0 +1,219 @@
+        }
+
+        [Fact]
+        public async Task Cannot_create_OneToOne_relationship_from_dependent_side()
```
This one can be removed.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Creating/CreateResourceWithRelationshipsTests.cs:
```diff
@@ -0,0 +1,219 @@
+        }
+
+        [Fact]
+        public async Task Cannot_create_relationship_with_include()
```
This one can be removed too. Creating a resource with a relationship is not possible, so doing that and then fetching the related data is not possible either.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Fetching/FetchResourceTests.cs:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        [Fact]
+        public async Task Cannot_get_primary_resource_for_unknown_type()
```
This is checked in the serializer and can be removed,

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Fetching/FetchResourceTests.cs:
```diff
@@ -0,0 +1,150 @@
+        }
+        
+        [Fact]
+        public async Task Cannot_get_secondary_HasOne_resource()
```
We should have a test for HasMany too, for example: GET /workItems/123/subscribers.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateResourceTests.cs:
```diff
@@ -0,0 +1,488 @@
+        }
+
+        [Fact]
+        public async Task Cannot_update_resource_for_missing_ID()
```
This is testing logic in the serializer, so can be removed. Basically, all errors containing "Failed to deserialize request body" can be taken out.

bart-degreed reviewed (changes requested):
Skimmed over the ReadWrite tests quickly, I'll take a closer look once you're done with them.

mrnkr commented:
I've just finished adding all the tests for `ReadWrite` that seemed relevant according to what you've told me in previous comments. I've also addressed all the feedback items so I requested another review round.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/WorkItemGroup.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.ReadWrite
+{
+    public class WorkItemGroup : MongoDbIdentifiable
```
The 'Items' relationship property was removed from this class. Probably because I've asked before to removed unused code. The difference here is that these models were thought out to be a logical/consistent set of various kinds of attributes and relationships going in various directions, to provide the tests everything they could possibly need. So I'd prefer to not cut out parts any parts here.

This does not apply to the models in the example project itself, which has grown over time and has become messy. We intend to clean that up in the future and aim for the same quality level as in the ReadWrite section.

The same applies to `WriteFakers`, which is supposed to provide a building block for the set of ReadWrite models. I wouldn't cut out parts of that either. (I now realize the name WriteFakers should have been ReadWriteFakers instead, but we forgot to rename it later in the project.)

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/WorkItem.cs:
```diff
@@ -0,0 +1,58 @@
+        public IList<WorkItem> RelatedFrom { get; set; }
+        public IList<WorkItemToWorkItem> RelatedFromItems { get; set; }
+
+        [BsonIgnore]
```
In this file, it looks as if you replaced all occurrences of `[NotMapped]` with `[BsonIgnore]`. I wonder what that means for the other relationships (for example: Parent), that don't have BsonIgnore on them. Shoudl they? I guess that now the related resources will be stored inside the same document, but I don't know enough of MongoDB to reason about that.

I'm not saying changes are needed, just wanted to make you aware of this. At other places where relationships are used, you did add BsonIgnore on them, so I don't see a consistent approach in how to deal with that.

mrnkr replied:
The only relationships in which not having `BsonIgnore` is problematic are those which are recursive, regular relationships should be stored as nested docs as you say. In order to avoid any confusion I'd thought of just putting `BsonIgnore` in all of them since they're just being used to test that we get the proper error when trying to reference relationships so here is just an example of me forgetting to do just that.

bart-degreed replied:
Thanks for the clarification.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Relationships/AddToToManyRelationshipTests.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+        
+        [Fact]
+        public async Task Can_add_to_HasManyThrough_relationship()
```
Cannot

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Relationships/AddToToManyRelationshipTests.cs:
```diff
@@ -0,0 +1,107 @@
+
+namespace JsonApiDotNetCoreMongoDbExampleTests.IntegrationTests.ReadWrite.Updating.Relationships
+{
+    public sealed class AddToToManyRelationshipTests
```
As mentioned earlier, I'd prefer to create the related resources in the database upfront for these tests, because they are being referenced in the request.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Relationships/UpdateToOneRelationshipTests.cs:
```diff
@@ -0,0 +1,91 @@
+        }
+        
+        [Fact]
+        public async Task Cannot_replace_OneToOne_relationship()
```
The distinction between OneToOne and ManyToOne is not relevant here, so pick one of these two tests, rename to Cannot_replace_relationship and remove the other one.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/ReplaceToManyRelationshipTests.cs:
```diff
@@ -0,0 +1,130 @@
+        public async Task Cannot_replace_HasMany_relationship()
+        {
+            // Arrange
+            var existingWorkItem = new WorkItem();
```
Aside from earlier remarks on faker usage, which applies to multiple files which I'm not all going to highlight, the fact that only a workitem (without relationships) is being created makes that the test name no longer matches the scenario. Instead of replacement, it has become assignment. I think it would be preferable to stick with the original intent.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateResourceTests.cs:
```diff
@@ -0,0 +1,416 @@
+        }
+
+        [Fact]
+        public async Task Can_completely_update_resource_with_string_ID()
```
There should also be a test for partial update.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateResourceTests.cs:
```diff
@@ -0,0 +1,416 @@
+        }
+
+        [Fact]
+        public async Task Can_completely_update_resource_with_string_ID()
```
Similar to CreateResourceTests, add a failing test when int is used for id.

mrnkr replied:
Should I add a new resource type and leave its Id as `string`? Hence, not inheriting from `MongoDbIdentifiable`?

bart-degreed replied:
If you mean `int` instead of `string`, then yes.

mrnkr replied:
That's what I meant, sorry.

I wrote an initial version of the test, not putting any particular thought into what the assertions were for the time being, and I could not get past the service locator trying to instantiate an `EntityFrameworkCoreRepository`... I tried explicitly registering the `MongoDbRepository<ModelWithIntId, int>` and still I get this error...

I'll upload this in my next commit and perhaps you can help me figure out what it is I did wrong...

bart-degreed replied:
I've pushed a commit to make that work.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateResourceTests.cs:
```diff
@@ -0,0 +1,416 @@
+            responseDocument.SingleData.Attributes["description"].Should().Be(newDescription);
+            responseDocument.SingleData.Attributes["dueAt"].Should().BeNull();
+            responseDocument.SingleData.Attributes["priority"].Should().Be(existingWorkItem.Priority.ToString("G"));
+            responseDocument.SingleData.Attributes.Should().ContainKey("concurrencyToken");
```
Relationships assertions are missing.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateResourceTests.cs:
```diff
@@ -0,0 +1,416 @@
+        }
+
+        [Fact]
+        public async Task Cannot_update_on_resource_ID_mismatch_between_url_and_body()
```
Can be removed.

bart-degreed commented on test/JsonApiDotNetCoreMongoDbExampleTests/IntegrationTests/ReadWrite/Updating/Resources/UpdateToOneRelationshipTests.cs:
```diff
@@ -0,0 +1,109 @@
+        }
+        
+        [Fact]
+        public async Task Cannot_create_OneToOne_relationship_from_principal_side()
```
Earlier remarks apply here too: test name, fakers, related resource creation. A single test is sufficient in this file.

bart-degreed reviewed (changes requested):
I've gone over all ReadWrite tests now.

mrnkr commented:
Addressed all comments but two for which I left a question

mrnkr commented:
Something else I forgot to mention, in adding the CreateResourceWithClientGeneratedIdTests class I saw I hadn't handled duplicate ID errors.

I saw in the main repo that exception is thrown from the service only and not from the repo, which I think throws a DataStoreUpdateException. Must have been my mistake but I couldn't make that work today... Did I overlook some key detail?

bart-degreed reviewed (changes requested):
Reviewed changes and added comments. Pushed commits to fix relationship links and duplicate IDs. The service catches `DataStoreUpdateException` when a write fails, so the repository must throw that. By using this exception, the service does not need to know about EF Core- or MongoDB-specific exception types.

bart-degreed reviewed (approved):
I think this is ready for merge now. Will hold off until I hear from you, in case I missed something. After merge, I intend to publish another alpha to NuGet.

mrnkr commented:
I'm just taking the liberty to rename `src/JsonApiDotNetCore.MongoDb/Configuration/public static class ServiceCollectionExtensions.cs` to `src/JsonApiDotNetCore.MongoDb/Configuration/ServiceCollectionExtensions.cs` hehe

Other than that I'm happy to have this merged :)

bart-degreed commented:
Ah, good catch!

Thanks a lot for your continued efforts working on this!
